File: features/transaction/purchase_platform/src/main/java/com/tokopedia/purchase_platform/features/cart/view/UpdateCartIntentService.java
package com.tokopedia.purchase_platform.features.cart.view;

import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.core.app.JobIntentService;

import com.google.gson.Gson;
import com.tokopedia.abstraction.base.app.BaseMainApplication;
import com.tokopedia.abstraction.common.utils.TKPDMapParam;
import com.tokopedia.abstraction.common.utils.network.AuthUtil;
import com.tokopedia.purchase_platform.features.cart.data.model.request.UpdateCartRequest;
import com.tokopedia.purchase_platform.features.cart.domain.model.cartlist.CartItemData;
import com.tokopedia.purchase_platform.features.cart.domain.model.cartlist.UpdateCartData;
import com.tokopedia.purchase_platform.features.cart.domain.usecase.UpdateCartUseCase;
import com.tokopedia.purchase_platform.features.cart.view.di.DaggerNewCartComponent;
import com.tokopedia.usecase.RequestParams;
import com.tokopedia.user.session.UserSessionInterface;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import rx.Subscriber;

/**
 * @author Irfan Khoirul on 31/07/18.
 */

public class UpdateCartIntentService extends JobIntentService {

    public static final String EXTRA_CART_ITEM_DATA_LIST = "EXTRA_CART_ITEM_DATA_LIST";

    @Inject
    UpdateCartUseCase updateCartUseCase;

    @Inject
    UserSessionInterface userSession;

    public UpdateCartIntentService() {
    }

    @Override
    public void onCreate() {
        super.onCreate();
        initInjector();
    }

    @Override
    protected void onHandleWork(@NonNull Intent intent) {
        if (userSession != null && updateCartUseCase != null && intent.hasExtra(EXTRA_CART_ITEM_DATA_LIST)) {
            List<CartItemData> cartItemDataList = intent.getParcelableArrayListExtra(EXTRA_CART_ITEM_DATA_LIST);
            List<UpdateCartRequest> updateCartRequestList = new ArrayList<>();
            for (CartItemData data : cartItemDataList) {
                updateCartRequestList.add(new UpdateCartRequest.Builder()
                        .cartId(data.getOriginData().getCartId())
                        .notes(data.getUpdatedData().getRemark())
                        .quantity(data.getUpdatedData().getQuantity())
                        .build());
            }
            TKPDMapParam<String, String> paramUpdate = new TKPDMapParam<>();
            paramUpdate.put(UpdateCartUseCase.PARAM_CARTS, new Gson().toJson(updateCartRequestList));

            RequestParams requestParams = RequestParams.create();
            requestParams.putObject(UpdateCartUseCase.PARAM_REQUEST_AUTH_MAP_STRING_UPDATE_CART,
                    getGeneratedAuthParamNetwork(paramUpdate));

            updateCartUseCase.execute(requestParams, new Subscriber<UpdateCartData>() {
                @Override
                public void onCompleted() {

                }

                @Override
                public void onError(Throwable e) {
                    e.printStackTrace();
                }

                @Override
                public void onNext(UpdateCartData updateCartData) {
                    // Expected to do nothing
                }
            });
        }
    }

    private void initInjector() {
        BaseMainApplication baseMainApplication = (BaseMainApplication) getApplication();
        DaggerNewCartComponent.builder()
                .baseAppComponent(baseMainApplication.getBaseAppComponent())
                .build()
                .inject(this);
    }

    private TKPDMapParam<String, String> getGeneratedAuthParamNetwork(
            TKPDMapParam<String, String> originParams) {
        return originParams == null
                ? AuthUtil.generateParamsNetwork(
                getBaseContext(), userSession.getUserId(),
                userSession.getDeviceId())
                : AuthUtil.generateParamsNetwork(
                getBaseContext(), originParams,
                userSession.getUserId(),
                userSession.getDeviceId()
        );
    }

}

File: features/transaction/purchase_platform/src/main/java/com/tokopedia/purchase_platform/features/cart/view/UpdateCartIntentService.kt
package com.tokopedia.purchase_platform.features.cart.view

import android.content.Intent
import androidx.core.app.JobIntentService
import com.tokopedia.abstraction.base.app.BaseMainApplication
import com.tokopedia.purchase_platform.features.cart.data.model.request.UpdateCartRequest
import com.tokopedia.purchase_platform.features.cart.domain.model.cartlist.CartItemData
import com.tokopedia.purchase_platform.features.cart.domain.model.cartlist.UpdateCartData
import com.tokopedia.purchase_platform.features.cart.domain.usecase.UpdateCartUseCase
import com.tokopedia.purchase_platform.features.cart.view.di.DaggerCartComponent
import com.tokopedia.usecase.RequestParams
import com.tokopedia.user.session.UserSessionInterface
import rx.Subscriber
import java.util.*
import javax.inject.Inject

/**
 * @author Irfan Khoirul on 31/07/18.
 */

class UpdateCartIntentService : JobIntentService() {

    @Inject
    lateinit var updateCartUseCase: UpdateCartUseCase
    @Inject
    lateinit var userSession: UserSessionInterface

    override fun onCreate() {
        super.onCreate()
        initInjector()
    }

    override fun onHandleWork(intent: Intent) {
        if (intent.hasExtra(EXTRA_CART_ITEM_DATA_LIST)) {
            val cartItemDataList = intent.getParcelableArrayListExtra<CartItemData>(EXTRA_CART_ITEM_DATA_LIST)
            val updateCartRequestList = getUpdateCartRequest(cartItemDataList)
            val requestParams = RequestParams.create()
            requestParams.putObject(UpdateCartUseCase.PARAM_UPDATE_CART_REQUEST, updateCartRequestList)

            updateCartUseCase.createObservable(requestParams)
                    .subscribe(object : Subscriber<UpdateCartData>() {
                        override fun onCompleted() {

                        }

                        override fun onError(e: Throwable) {
                            e.printStackTrace()
                        }

                        override fun onNext(updateCartData: UpdateCartData) {
                            // Expected to do nothing
                        }
                    })
        }
    }

    private fun getUpdateCartRequest(cartItemDataList: List<CartItemData>): ArrayList<UpdateCartRequest> {
        val updateCartRequestList = ArrayList<UpdateCartRequest>()
        for (cartItemData in cartItemDataList) {
            val updateCartRequest = UpdateCartRequest()
            updateCartRequest.cartId = cartItemData.originData?.cartId?.toString() ?: "0"
            updateCartRequest.notes = cartItemData.updatedData?.remark
            updateCartRequest.quantity = cartItemData.updatedData?.quantity ?: 0
            updateCartRequestList.add(updateCartRequest)
        }
        return updateCartRequestList
    }

    private fun initInjector() {
        val baseMainApplication = application as BaseMainApplication
        DaggerCartComponent.builder()
                .baseAppComponent(baseMainApplication.baseAppComponent)
                .build()
                .inject(this)
    }

    companion object {
        val EXTRA_CART_ITEM_DATA_LIST = "EXTRA_CART_ITEM_DATA_LIST"
    }

}
